### DISCUSSION TITLE
Documentation about algorithms

### DISCUSSION THREAD

**Discussion by ming-tsai** [MakeContributions/DSA#206] [Ideas]
Hi guys,
Now a day we have all algorithms, but we don't have the documentation about each one of them, I think to make some changes to the folder.
- Move all algorithms to the algorithms folder
- Create a docs folder, it will contain the docs of the algorithms

What are you thinking?

**Comment by TawfikYasser** [comment]
Yes @ming-tsai It is a good idea, Really, I feel that the repository is now disorganized and needs to be indexed to be easier for learners and also for new contributors.

**Comment by TawfikYasser** [comment]
Daily I'm thinking about how to make the repo. more attractive and to not only contain DSA

We want to expand the fields in order to provide the opportunity for a larger number of people to contribute, for example adding other types of contributions, not only algorithms and data structures (DSA), for example, another part for beginners in programming and CS in general to add the small programs that they work on on the one hand they learn how to deal with GitHub and from On the other hand, in preserving their works, and so with other types, what do you think @ming-tsai , and we can also see suggestions.

**Comment by ming-tsai** [comment]
Good point

**Comment by cclauss** [comment]
In building https://github.com/TheAlgorithms we took a _one-repository-per-language_ approach and separate  [Algorithms Explanation](https://github.com/TheAlgorithms/Algorithms-Explanation) repo.  repo-per-language allows contributors and maintainers to focus on the languages that they know (or want to learn).  Communities build separately but there is a lot of compare and contrast that maintainers do to move best practices from one language community to another.  Automated testing usually varies quite a bit from language to language so the community can also collaborate to demonstrate best practices for automated testing and lining for their focused language.  Moving the algorithm explanations to a separate repo allows doc-centric contributors to build the docs in multiple human languages.  It ensures that the docs are not built too close to only one implementation.  Another trick that is worth looking at is that most repos _automatically_ build a `DIRECTORY.md` file as new algorithms are added.  Without automation, this becomes an increasingly error-prone tedious task.

**Comment by apex-blaze** [comment]
Yes, I actually agree with @cclauss , because the file structure we currently have is kinda mess. We have folders denoting the data structures, within them we have sub-classified our languages. Due to this, alot of algorithms have been coded in just 1 or 2 languages and rest are missing. It's better to have languages separated from root folder or the one @cclauss  has mentioned.

**Comment by ming-tsai** [comment]
Actually, I would like to make some different way of **TheAlgorithms**, I'm thinking we could be as below folder structure:
```
algorithms
  |--c-or-c++
     |-- search
         |-- binary-search.cpp
     |-- README
  |--java
     |-- search
        |-- binary-search.java
     |-- README
  |--csharp
     |-- search
         |-- binary-search.cs
     |-- README
  |--python
     |-- search
         |-- binary-search.py
     |-- README
docs
  |-- search
     |-- binary-search.md

```
For each algorithm, the folder contains the source and tests, then we could create the Github action to check each language, and create an action that could update the index readme automatically like this [one](https://github.com/MakeContributions/ideahub/actions/workflows/scraper.yml)

**Comment by ming-tsai** [comment]
This is my opinion, and we won't need to change all files, only need to move folder, create the actions and create automatically job

**Comment by cclauss** [comment]
That works.

My sense is that languages are brand names (I bet language communities agree) so C, C++, JavaScript, Python are more respectful than all lowercase.  python is a snake and Python is a programming language.

**Comment by ming-tsai** [comment]
Ok, we could change that rule

**Comment by ming-tsai** [comment]
I will starting moving CSharp and Javascript folder

**Comment by ming-tsai** [comment]
Done with [C#](https://github.com/MakeContributions/DSA/tree/main/algorithms/CSharp) [![.NET](https://github.com/MakeContributions/DSA/actions/workflows/dotnet.yml/badge.svg)](https://github.com/MakeContributions/DSA/actions/workflows/dotnet.yml)

**Comment by cclauss** [comment]
JavaScript... Capital J and capital S

**Comment by UG-SEP** [comment]
> In building https://github.com/TheAlgorithms we took a _one-repository-per-language_ approach and separate [Algorithms Explanation](https://github.com/TheAlgorithms/Algorithms-Explanation) repo. repo-per-language allows contributors and maintainers to focus on the languages that they know (or want to learn). Communities build separately but there is a lot of compare and contrast that maintainers do to move best practices from one language community to another. Automated testing usually varies quite a bit from language to language so the community can also collaborate to demonstrate best practices for automated testing and lining for their focused language. Moving the algorithm explanations to a separate repo allows doc-centric contributors to build the docs in multiple human languages. It ensures that the docs are not built too close to only one implementation. Another trick that is worth looking at is that most repos _automatically_ build a `DIRECTORY.md` file as new algorithms are added. Without automation, this becomes an increasingly error-prone tedious task.



> In building https://github.com/TheAlgorithms we took a _one-repository-per-language_ approach and separate [Algorithms Explanation](https://github.com/TheAlgorithms/Algorithms-Explanation) repo. repo-per-language allows contributors and maintainers to focus on the languages that they know (or want to learn). Communities build separately but there is a lot of compare and contrast that maintainers do to move best practices from one language community to another. Automated testing usually varies quite a bit from language to language so the community can also collaborate to demonstrate best practices for automated testing and lining for their focused language. Moving the algorithm explanations to a separate repo allows doc-centric contributors to build the docs in multiple human languages. It ensures that the docs are not built too close to only one implementation. Another trick that is worth looking at is that most repos _automatically_ build a `DIRECTORY.md` file as new algorithms are added. Without automation, this becomes an increasingly error-prone tedious task.

agree 💯💯

**Comment by ming-tsai** [comment]
> JavaScript... Capital J and capital S

I'm going to working to that one

**Comment by ming-tsai** [comment]
We could specify the paths for the project like this [action](https://github.com/MakeContributions/DSA/blob/main/.github/workflows/dotnet.yml)

**Comment by ming-tsai** [comment]
Hi guys,
Someone could help me with C++, C, Go, Java, and Python projects?
@Arsenic-ATG, @cclauss, @TawfikYasser, @UG-SEP, @atin, @aayushjain7

**Reply by UG-SEP** [reply]
yes i can in c

**Reply by ming-tsai** [reply]
Ok, I will wait your pull request @UG-SEP

**Reply by UG-SEP** [reply]
> Ok, I will wait your pull request @UG-SEP

Soon I will do it

**Comment by TawfikYasser** [comment]
Hi @ming-tsai 
I'm here, What is the type of work should I do?

**Reply by ming-tsai** [reply]
See below list:
- Move the language file to their folder
- Create a project for the language
- Create a test project (optional)
- Create a workflow for it

Example [C#](https://github.com/MakeContributions/DSA/tree/main/algorithms/CSharp) and [JS](https://github.com/MakeContributions/DSA/tree/enhacement/moving-js-to-project/algorithms/JS)

**Reply by ming-tsai** [reply]
I'm working with JS

**Reply by TawfikYasser** [reply]
Okay I'll work on Java, Thanks @ming-tsai for the clearfiy.

**Reply by TawfikYasser** [reply]
Please give me time because I've deliverers in my college. I will work on **Java**

**Reply by ming-tsai** [reply]
Ok, take it easy

**Reply by TawfikYasser** [reply]
Hi @ming-tsai , Java is Done #220

**Comment by cclauss** [comment]
I could take Python if that would help.

**Comment by Arsenic-ATG** [comment]
I am up for taking care of C++ programs, and I think this is the right time to separate C programs from C++ programs instead of putting them at one place.

**Reply by UG-SEP** [reply]
> I am up for taking care of C++ programs, and I think this is the right time to separate C programs from C++ programs instead of putting them at one place.

yes

**Reply by ming-tsai** [reply]
Nice, I will wait for your pull request

**Comment by ming-tsai** [comment]
So, we only need finished C, C++, and Go, then starting the documentation and update de `README.md`

**Comment by ming-tsai** [comment]
Hi guys,
Now we could start we the documentation of all algorithms please see this [link](https://github.com/MakeContributions/DSA/tree/main/docs/en)

**Reply by TawfikYasser** [reply]
Okay @ming-tsai  I'll start soon
